### PR TITLE
Add a new archiver client which doesn't retry

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added a new archival tool which doesn't rely on the archival workflow.

**Why?**
I made this change so that we can deprecate the archival workflow and move to the archival queue.

**How did you test it?**
I tested this change by adding unit tests which verify the side-effects, metrics, logs and return errors for the archiver with different input targets and errors.

**Potential risks**
Nothing--it's not used in production.

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/worker/archiver/client_v2.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		sdkClientFactory sdk.ClientFactory
+		numWorkflows     dynamicconfig.IntPropertyFn
+		signalTimeout    dynamicconfig.DurationPropertyFn
+		rateLimiter      quotas.RateLimiter
+		archiverProvider provider.ArchiverProvider
+	}
+)
+
+// NewClientV2 creates a new ClientV2
+func NewClientV2(
```
Comment (yux0) [inline service/worker/archiver/client_v2.go:103]: should the client locate in the history service? If so, it should remove the v2  in the name.

Comment (MichaelSnowden) [reply service/worker/archiver/client_v2.go]: Moved into history/archival

File: service/history/archival/service.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	// Client, which will try to signal an archival workflow whenever an error occurrs.
+	Service interface {
+		Archive(context.Context, *Request) (*Response, error)
+	}
+
+	service struct {
+		metricsScope     metrics.Scope
+		logger           log.Logger
+		sdkClientFactory sdk.ClientFactory
+		numWorkflows     dynamicconfig.IntPropertyFn
```
Comment (yux0) [inline service/history/archival/service.go:97]: not used

File: service/history/archival/service.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	// Service is used to archive workflow history and visibility data. If a target fails, it returns an error, unlike
+	// Client, which will try to signal an archival workflow whenever an error occurrs.
+	Service interface {
+		Archive(context.Context, *Request) (*Response, error)
+	}
+
+	service struct {
+		metricsScope     metrics.Scope
+		logger           log.Logger
+		sdkClientFactory sdk.ClientFactory
```
Comment (yux0) [inline service/history/archival/service.go:96]: not used

File: service/history/archival/service.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		numWorkflows     dynamicconfig.IntPropertyFn
+		signalTimeout    dynamicconfig.DurationPropertyFn
```
Comment (yux0) [inline service/history/archival/service.go:98]: not used

File: service/history/archival/service.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		rateLimiter      quotas.RateLimiter
```
Comment (yux0) [inline service/history/archival/service.go:99]: not used

File: service/history/archival/service.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+		}
+	}
+	wg.Wait()
+	return &Response{}, multierr.Combine(errs...)
+}
+
+func (c *service) archiveHistory(ctx context.Context, request *Request, logger log.Logger) (err error) {
+	logger = log.With(
+		logger,
+		tag.ShardID(request.ShardID),
```
Comment (yux0) [inline service/history/archival/service.go:178]: this is declared in L144

File: service/history/archival/service.go (hunk 5 above)
Comment (yux0) [reply service/history/archival/service.go:99]: but I agree we should add the rate limit for archival provider

File: service/history/archival/service.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+) Service {
+	return &service{
+		archiverProvider: archiverProvider,
+		metricsScope:     metricsClient.Scope(metrics.ArchiverClientScope),
+		logger:           logger,
+		rateLimiter:      rateLimiter,
+	}
+}
+
+type RateLimitExceededErr struct {
```
Comment (yux0) [inline service/history/archival/service.go:120]: Not need to have customized error. Use serviceerror.ResourceExhausted error

File: service/history/archival/service.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	Target string
+
+	Response struct {
+	}
+
+	// Service is used to archive workflow history and visibility data. If a target fails, it returns an error, unlike
+	// Client, which will try to signal an archival workflow whenever an error occurrs.
+	Service interface {
```
Comment (yux0) [inline service/history/archival/service.go:88]: nit: we have four services(frontend, history matching, worker). What about just keep it as archiver?

**Review by yux0 — APPROVED** [review]

**Pull request merged**